src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1987,6 +2154,57 @@ def translate_device(dev):
         else:
             return False
 
+    @private
+    def get_iommu_type(self):
+        async def check_util_output(proc, pattern):
+            found = False
+            while True:
+                line = await proc.stdout.readline()
+                if line:
+                    if re.search(pattern, line.decode('utf-8')):
+                        found = True
+                else:
+                    break  # Reached EOF
+            return found
+
+        async def ext_util(iommu_type, cmd_args, pattern):
+            found = False
+            proc = None
+            try:
+                proc = await asyncio.create_subprocess_exec(*cmd_args,
```

Comment:
So we already have this in place and we can just reuse the existing implementation reducing duplication and making it more neat ; )
Please see this example - https://github.com/freenas/freenas/blob/66c51a77288731d3eaeca3a2dc756c4d0e12669b/src/middlewared/middlewared/plugins/disk_/swap_configure.py#L78

You can use this for subprocess calls

Reply:
If I understand correctly `middlewared.utils.run` will not return until the subprocess has finished and the complete output will be buffered in memory before it becomes available to the caller. Probably ok if the subprocess is well bahaved and does not produce too much output, but in case it misbehaves could it not potentially make middlewared unstable or run out of memory?

My intention was to avoid the need to buffer much in python and to avoid getting stuck by doing asynchronous readlines and by terminating the subprocess after a timeout if it does not complete within reasonable time. A bit safer I think although, admittedly, some more code. What do you think?

Reply:
I just saw there is similar handling in this example:
https://github.com/freenas/freenas/blob/66c51a77288731d3eaeca3a2dc756c4d0e12669b/src/middlewared/middlewared/plugins/disk_/hddstandby_force.py#L58-L88
albeit the scope of the timeout timer is slightly different (restarted for each line read instead of only before the first line).

Reply:
@sonicaj, unless you have a strong preference I think it would be prudent to keep the extra protection. The PR has been updated (less this aspect) and is ready for review. 

Thank you for constructive feedback and suggestions!

Reply:
See reply to same question in other comment.